src/libpriv/rpmostree-core.c

```diff
@@ -2393,12 +2424,151 @@ canonicalize_non_usrmove_path (RpmOstreeContext  *self,
+    }
+
+  /* this we *do* canonicalize since we'll be comparing against ostree paths */
+  g_autoptr(GHashTable) files_skip_add =
+    g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  /* this one we *don't* canonicalize since we'll be comparing against rpmfi paths */
+  g_autoptr(GHashTable) files_skip_delete =
+    g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+
+  /* we deal with color similarly to librpm (compare with skipInstallFiles()) */
+  rpm_color_t ts_color = rpmtsColor (ts);
+  rpm_color_t ts_prefcolor = rpmtsPrefColor (ts);
+
+  /* ignore colored files not in our rainbow */
+  GLNX_HASH_TABLE_FOREACH_IT (files_added, it, const char*, fn, gpointer, colorp)
+    {
+      rpm_color_t color = GPOINTER_TO_UINT (colorp);
+      if (color && ts_color && !(ts_color & color))
+        g_hash_table_add (files_skip_add, canonicalize_rpmfi_path (fn));
+    }
+
+  g_auto(rpmdbMatchIterator) it = rpmtsInitIterator (ts, RPMDBI_PACKAGES, NULL, 0);
+  Header h;
+  while ((h = rpmdbNextIterator (it)) != NULL)
+    {
+      /* is this pkg to be deleted? */
+      guint off = rpmdbGetIteratorOffset (it);
+      if (g_hash_table_contains (pkgs_deleted, GUINT_TO_POINTER (off)))
+        continue;
+
+      /* try to only load what we need: filenames and colors */
+      rpmfiFlags flags = RPMFI_FLAGS_QUERY;
+#ifdef BUILDOPT_HAVE_RPMFILES /* we're using this more as a "if not on CentOS" switch */
+      /* this is even more restrictive that QUERY */
+      flags = RPMFI_FLAGS_ONLY_FILENAMES;
+#endif
+      flags &= ~RPMFI_NOFILECOLORS;
+
+      g_auto(rpmfi) fi = rpmfiNew (ts, h, RPMTAG_BASENAMES, flags);
+
+      fi = rpmfiInit (fi, 0);
+      while (rpmfiNext (fi) >= 0)
+        {
+          const char *fn = rpmfiFN (fi);
+          g_assert (fn != NULL);
+
+          /* check if one of the pkgs to delete wants to delete our file */
+          if (g_hash_table_contains (files_deleted, fn))
+            g_hash_table_add (files_skip_delete, g_strdup (fn));
+
+          rpm_color_t color = (rpmfiFColor (fi) & ts_color);
+
+          /* let's make the safe assumption that the color mess is only an issue for /usr */
+          const char *fn_rel = fn;
+          fn_rel += strspn (fn, "/");
+
+          /* be sure we've canonicalized usr/ */
+          g_autofree char *fn_rel_owned = canonicalize_non_usrmove_path (self, fn_rel);
+          if (fn_rel_owned)
+            fn_rel = fn_rel_owned;
+
+          if (!g_str_has_prefix (fn_rel, "usr/"))
+            continue;
+
+          /* check if one of the pkgs to install wants to overwrite our file */
+          rpm_color_t other_color =
+            GPOINTER_TO_UINT (g_hash_table_lookup (files_added, fn));
+          other_color &= ts_color;
+
+          /* see handleColorConflict() */
+          if (color && other_color && (color != other_color))
+            {
+              /* do we already have the preferred color installed? */
+              if (color & ts_prefcolor)
```

Comment:
Just for my own understanding...could this be `==` instead of `&`?

Reply:
Yes, you're correct AFAIU. They're both bitfields, but `ts_color` contains all the colors we're compatible with, while `ts_prefcolor` contains only a single color from that set. And files can only one specific color, so `==` would work equally well here.